### PR TITLE
Support magic values prefixing types

### PR DESCRIPTION
**Pull request by samuelsleight**
This PR as-is is an initial implementation of #109, hacked-in with very little experience in this library or with developing proc-macros in general.

It currently supports reads only, and is very specific to my use-case, so will want to be expanded to support writes and to be more useful in general, and will also want tests implementing and documentation adding for the new attribute.

### CODE REVIEW

File: deku-derive/src/macros/deku_read.rs [hunk 1]
```diff
@@ -38,6 +38,23 @@ fn emit_struct(input: &DekuData) -> Result<TokenStream, syn::Error> {
 
     let initialize_struct = super::gen_struct_init(is_named_struct, internal_fields);
 
+    let check_magic = if let Some(magic) = &input.magic {
```
Comment (sharksforarms) [inline deku-derive/src/macros/deku_read.rs:41]: Similarly, check_magic should also be available for [emit_enum](https://github.com/sharksforarms/deku/blob/22e271d421574cf9ebc6f9e00c42feb0978ecb70/deku-derive/src/macros/deku_read.rs#L249)

File: deku-derive/src/lib.rs [hunk 2]
```diff
@@ -29,6 +29,9 @@ struct DekuData {
     /// default context passed to the field
     ctx_default: Option<Punctuated<syn::Expr, syn::token::Comma>>,
 
+    /// A magic value that must appear at the start of this struct's data
```
Comment (sharksforarms) [inline deku-derive/src/lib.rs:32]: `A magic value that must appear at the start` (for both structs/enums)

File: deku-derive/src/lib.rs [hunk 3]
```diff
@@ -386,6 +390,10 @@ struct DekuReceiver {
     #[darling(default)]
     ctx_default: Option<syn::LitStr>,
 
+    /// A magic value that must appear at the start of this struct's data
```
Comment (sharksforarms) [inline deku-derive/src/lib.rs:393]: `A magic value that must appear at the start` (for both structs/enums)

**Review by sharksforarms — COMMENTED** [review]
Looks good so far! Couple of comments
- A function could be created such as `emit_magic` in the same file to share the "magic reading" tokens between `emit_struct` and `emit_enum`
- A success/failure test case should be added here: https://github.com/sharksforarms/deku/tree/master/tests/test_attributes
- Documentation of the new attribute should be added here: https://github.com/sharksforarms/deku/blob/master/src/attributes.rs#L6
- deku_write.rs write implementation, layout is similar to deku_read.rs (emit_struct/emit_enum) (write magic value before writing fields): https://github.com/sharksforarms/deku/blob/master/deku-derive/src/macros/deku_write.rs

**Comment by samuelsleight** [comment]
Thanks for the pointers! I'll hopefully get a chance to look into these tomorrow

**Comment by samuelsleight** [comment]
Cool, this should now be somewhat less hacked-in!

File: src/attributes.rs [hunk 4]
```diff
@@ -101,6 +102,36 @@ let value: Vec<u8> = value.try_into().unwrap();
...
+that type's data when writing, without needing to be explicitly stored
+in a field
+
+Example:
+```rust
+# use deku::prelude::*;
+# use std::convert::{TryInto, TryFrom};
+# #[derive(Debug, PartialEq, DekuRead, DekuWrite)]
+#[deku(magic = b"deku")]
+pub struct MagicDeku {
```
Comment (sharksforarms) [inline src/attributes.rs:118]: nit: `DekuTest` is used as the struct names in the docs for consistency between examples

File: tests/test_magic.rs [hunk 5]
```diff
@@ -0,0 +1,105 @@
+use deku::prelude::*;
+use hexlit::hex;
+use rstest::rstest;
+use std::convert::{TryFrom, TryInto};
+
+pub mod samples {
```
Comment (sharksforarms) [inline tests/test_magic.rs:6]: I got rid of this `samples` pattern during a recent refactor, you can place the structs/enums inside the test functions :) Can check `test_cond.rs` for an example

File: tests/test_magic.rs [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+    #[deku(magic = b"UKED")]
+    pub struct NestedMagicDeku {
+        pub nested: MagicDeku,
+    }
+}
+
+#[rstest(input,
+    case(&hex!("64656b75")),
+
+    #[should_panic(expected = "Parse(\"Missing magic value [100, 101, 107, 117]\")")]
```
Comment (sharksforarms) [inline tests/test_magic.rs:30]: Are the multiple "Missing magic value" tests testing different code paths?

File: tests/test_magic.rs (hunk 5 above)
Comment (sharksforarms) [reply tests/test_magic.rs:6]: Also used `TestStruct` or `TestEnum` inside the test functions for consistency

File: tests/test_magic.rs [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+#[rstest(input,
+    case(&hex!("554b454464656b75")),
+
+    #[should_panic(expected = "Parse(\"Missing magic value [100, 101, 107, 117]\")")]
+    case(&hex!("554b4544deadbeef")),
+
+    #[should_panic(expected = "Parse(\"Missing magic value [85, 75, 69, 68]\")")]
+    case(&hex!("deadbeef64656b75")),
+)]
+fn test_nested_magic_struct(input: &[u8]) {
```
Comment (sharksforarms) [inline tests/test_magic.rs:93]: I don't think this test case is necessary, there's a test case for nested cases in `test_struct.rs`

File: src/attributes.rs [hunk 8]
```diff
@@ -101,6 +102,36 @@ let value: Vec<u8> = value.try_into().unwrap();
 assert_eq!(data, value);
 ```
 
+# magic
+
+Sets a "magic" value that must be present in the data at the start of
+a struct/enum when reading, and that is written out of the start of
+that type's data when writing, without needing to be explicitly stored
```
Comment (sharksforarms) [inline src/attributes.rs:109]: `, without needing to be explicitly stored in a field` can be removed I think

**Review by sharksforarms — COMMENTED** [review]
Looks great! Some small comments.

File: tests/test_magic.rs (hunk 5 above)
Comment (samuelsleight) [reply tests/test_magic.rs:6]: Oh, yep, I missed that that refactor had removed the `samples` pattern

File: tests/test_magic.rs (hunk 6 above)
Comment (samuelsleight) [reply tests/test_magic.rs:30]: They are each testing a different byte of the input being incorrect, which I figured was worth checking for to future-proof against really bad implementations of magic reading that, say, only check for the last byte and assume the first few are correct.

(Not saying that specific case is going to happen, obviously, but I just figured it was better to be safe than sorry)

**Comment by sharksforarms** [comment]
Thanks!

**Pull request merged**